api_call_options.go

```diff
@@ -0,0 +1,162 @@
+	// behavior matches either a name or a fingerprint field, while another
+	// resource may match entirely different fields.  Search is currently
+	// implemented for SSHKeys and uses an exact match.
+	Search string `url:"search,omitempty"`
+
+	SortBy        string            `url:"sort_by,omitempty"`
+	SortDirection ListSortDirection `url:"sort_direction,omitempty"`
+
+	Meta meta `url:"-"`
+}
+
+type ListOptions = GetOptions
+type SearchOptions = GetOptions
+
+type QueryAppender interface {
+	WithQuery(path string) string // we use this in all List functions (urlQuery)
+	GetPage() int                 // we use this in List
+	Including(...string)          // we use this in Device List to add facility
+}
+
+// GetOptions returns GetOptions from GetOptions (and is nil-receiver safe)
+func (g *GetOptions) GetOptions() *GetOptions {
+	getOpts := GetOptions{}
+	if g != nil {
+		getOpts.Includes = g.Includes
+		getOpts.Excludes = g.Excludes
+	}
+	return &getOpts
+}
+
+func (g *GetOptions) WithQuery(path string) string {
+	params := g.Encode()
+	if params != "" {
+		// parse path, take existing vars
+		path = fmt.Sprintf("%s?%s", path, params)
+	}
+	return path
+}
+
+// OptionsGetter provides GetOptions
+type OptionsGetter interface {
+	GetOptions() *GetOptions
+}
+
+func (g *GetOptions) GetPage() int { // guaranteed int
+	if g == nil {
+		return 0
+	}
+	return g.Page
+}
+
+func (g *GetOptions) CopyOrNew() *GetOptions {
+	if g == nil {
+		return &GetOptions{}
+	}
+	ret := *g
+	return &ret
+}
+
+// Including ensures that the variadic refs are included in a copy of the
+// options, resulting in expansion of the the referred sub-resources. Unknown
+// values within refs will be silently ignore by the API.
+func (g *GetOptions) Including(refs ...string) *GetOptions {
+	ret := g.CopyOrNew()
+	for _, v := range refs {
+		if !contains(ret.Includes, v) {
+			ret.Includes = append(ret.Includes, v)
+		}
+	}
+	return ret
+}
+
+func stripQuery(inURL string) string {
+	u, _ := url.Parse(inURL)
+	u.RawQuery = ""
+	return u.String()
+}
+
+// nextPage is common and extracted from all List functions
+func nextPage(meta meta, opts *GetOptions) (path string) {
+	if meta.Next != nil && (opts.GetPage() == 0) {
+		optsCopy := opts.CopyOrNew()
+		optsCopy.Page = meta.CurrentPageNum + 1
+		return optsCopy.WithQuery(stripQuery(meta.Next.Href))
+	}
+	if opts != nil {
+		opts.Meta = meta
+	}
+	return ""
+}
+
+// Encode generates a URL query string ("?foo=bar")
+func (o *GetOptions) Encode() string {
+	urlValues, _ := query.Values(o)
+	return urlValues.Encode()
+}
+
+func urlQuery(o *GetOptions) string {
+	urlValues, _ := query.Values(o)
+	return urlValues.Encode()
+}
+
+// getApiCallUrl will construct URL path with variables. It will build the
+// URL path from segments, and encode the options. If there are no options
+// in the API call, pass nil for opts.
+func getApiCallUrl(opts *GetOptions, pathSegments ...string) string {
```

Comment:
@displague I added this whapper to further simplify the API call URL assembly in the packngo functions.

Reply:
replied in https://github.com/packethost/packngo/pull/211#discussion_r526347547